### PR TITLE
[shardformer/sequence parallel] Support sequence parallel for gpt2

### LINKED ISSUE #4383: [shardformer/sequence parallel] support sequence parallel for gpt2

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
fixed #4383 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
Support sequence parallel for gpt2

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/_operation.py [hunk 1]
```diff
@@ -141,6 +143,173 @@ def backward(ctx, grad_output):
...
+    """
+
+    @staticmethod
+    def forward(ctx, input_, process_group, dim):
+        ctx.dim = dim
+        ctx.process_group = process_group
+
+        # do reduce-scatter
+        new_shape = list(input_.shape)
+        new_shape[dim] = new_shape[dim] // dist.get_world_size(process_group)
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:230]: What if new shape is not dividable by world size?

File: colossalai/shardformer/modeling/gpt2_seq.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+from transformers.modeling_outputs import BaseModelOutputWithPastAndCrossAttentions
+from transformers.utils import logging
+
+from colossalai.shardformer.layer._operation import gather_forward_split_backward, split_forward_gather_backward
+from colossalai.shardformer.shard import ShardConfig
+
+logger = logging.get_logger(__name__)
+
+
+def seq_forward_fn(shard_config: ShardConfig):
```
Comment (ver217) [inline colossalai/shardformer/modeling/gpt2_seq.py:17]: This naming shoud be clearer.

File: tests/test_shardformer/test_model/test_shard_gpt2.py [hunk 3]
```diff
@@ -53,7 +53,7 @@ def check_forward_backward(model_fn, data_gen_fn, output_transform_fn, loss_fn,
         sharded_gpt2 = sharded_model.unwrap().transformer
 
     col_layer_for_check = ['h[0].mlp.c_fc']
-    row_layer_for_check = ['wte', 'h[0].mlp.c_proj']
+    row_layer_for_check = ['h[0].mlp.c_proj', 'wte']
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_shard_gpt2.py:56]: Does the order matter?

File: colossalai/shardformer/layer/_operation.py (hunk 1 above)
Comment (FoolPlayer) [reply colossalai/shardformer/layer/_operation.py:230]: Add assert before divide

File: colossalai/shardformer/modeling/gpt2_seq.py (hunk 2 above)
Comment (FoolPlayer) [reply colossalai/shardformer/modeling/gpt2_seq.py:17]: rename as `gpt2_sequence_parallel_forward_fn`

**Review by ver217 — APPROVED** [review]

**Pull request merged**